File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        lo, hi = 0, len(s) - 1

        while lo < hi:
            if not s[lo].isalnum(): lo += 1
            elif not s[hi].isalnum(): hi -= 1
            else:
                if s[lo].lower() != s[hi].lower(): return False
                else:
                    lo += 1
                    hi -= 1

        return True

File: 0125-valid-palindrome/0125-valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        int lo = 0;
        int hi = s.length() - 1;
        int n = s.length();

        while (lo < hi) {
            while (lo < n && !Character.isLetterOrDigit(s.charAt(lo))) {
                lo++;
            }
            while (hi > -1 && !Character.isLetterOrDigit(s.charAt(hi))) {
                hi--;
            }
            if (lo > hi) break;
            if (Character.toLowerCase(s.charAt(lo)) !=
                Character.toLowerCase(s.charAt(hi))) {
                    return false;
                }
            
            lo++;
            hi--;
        }

        return true;
    }
}